src/main/java/bio/terra/workspace/service/job/JobService.java

```diff
@@ -0,0 +1,381 @@
+   *   <li>Failed flight: if there is an exception, throw it. Note that we can only throw
+   *       RuntimeExceptions to be handled by the global exception handler. Non-runtime exceptions
+   *       require throw clauses on the controller methods; those are not present in the
+   *       swagger-generated code, so it introduces a mismatch. Instead, in this code if the caught
+   *       exception is not a runtime exception, then we throw JobResponseException passing in the
+   *       Throwable to the exception. In the global exception handler, we retrieve the Throwable
+   *       and use the error text from that in the error model
+   *   <li>Failed flight: no exception present. We throw InvalidResultState exception
+   * </ol>
+   *
+   * @param jobId to process
+   * @return object of the result class pulled from the result map
+   */
+  public <T> JobResultWithStatus<T> retrieveJobResult(
+      String jobId, Class<T> resultClass, AuthenticatedUserRequest userReq) {
+    boolean canListAnyJob = checkUserCanListAnyJob(userReq);
+
+    try {
+      // if the user has access to all jobs, then fetch the requested result
+      // otherwise, check that the user has access to it first
+      if (!canListAnyJob) {
+        verifyUserAccess(jobId, userReq); // jobId=flightId
+      }
+      return retrieveJobResultWorker(jobId, resultClass);
+    } catch (StairwayException stairwayEx) {
+      throw new InternalStairwayException(stairwayEx);
+    }
+  }
+
+  private <T> JobResultWithStatus<T> retrieveJobResultWorker(String jobId, Class<T> resultClass)
+      throws StairwayException {
+    FlightState flightState = stairway.getFlightState(jobId);
+    FlightMap resultMap = flightState.getResultMap().orElse(null);
+    if (resultMap == null) {
+      throw new InvalidResultStateException("No result map returned from flight");
+    }
+
+    switch (flightState.getFlightStatus()) {
+      case FATAL:
+      case ERROR:
+        if (flightState.getException().isPresent()) {
+          Exception exception = flightState.getException().get();
+          if (exception instanceof RuntimeException) {
+            throw (RuntimeException) exception;
+          } else {
+            throw new JobResponseException("wrap non-runtime exception", exception);
+          }
+        }
+        throw new InvalidResultStateException("Failed operation with no exception reported");
+
+      case SUCCESS:
+        HttpStatus statusCode =
+            resultMap.get(JobMapKeys.STATUS_CODE.getKeyName(), HttpStatus.class);
+        if (statusCode == null) {
+          statusCode = HttpStatus.OK;
+        }
+        return new JobResultWithStatus<T>()
+            .statusCode(statusCode)
+            .result(resultMap.get(JobMapKeys.RESPONSE.getKeyName(), resultClass));
+
+      case RUNNING:
+        throw new JobNotCompleteException(
+            "Attempt to retrieve job result before job is complete; job id: "
+                + flightState.getFlightId());
```

Comment:
hmm ... should this throw an Exception, or return a running status (maybe HTTP 100?) and a null result? Again, this might just be a style question.

Reply:
I kind of like leaving it as an exception - this way the call always returns either a result or an error message. That also makes it clear that this isn't the endpoint to poll, it's the one to get results from.